INDEX OBJECTS AND LABELED DATA

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("pokemon.csv")

In [3]:
data= data.set_index("#")

In [4]:
print(data.index.name)

#


.index.name metodu ile indeksin ismini görüntüledik.

In [5]:
data.index.name = "index_name"
data.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
index_name                                                               
1               Bulbasaur  Grass  Poison  45      49       49       65   
2                 Ivysaur  Grass  Poison  60      62       63       80   
3                Venusaur  Grass  Poison  80      82       83      100   
4           Mega Venusaur  Grass  Poison  80     100      123      122   
5              Charmander   Fire     NaN  39      52       43       60   

            Sp. Def  Speed  Generation  Legendary  
index_name                                         
1                65     45           1      False  
2                80     60           1      False  
3               100     80           1      False  
4               120     80           1      False  
5                50     65           1      False

İndeks ismini değiştirerek index_name yaptık.

In [6]:
data3 = data.copy()  #Burada orjinal data'nın bozulmasını önlemek için data'yı data3'e kopyaladım.
data3.index = range(100,900,1)
data3.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
100      Bulbasaur  Grass  Poison  45      49       49       65       65   
101        Ivysaur  Grass  Poison  60      62       63       80       80   
102       Venusaur  Grass  Poison  80      82       83      100      100   
103  Mega Venusaur  Grass  Poison  80     100      123      122      120   
104     Charmander   Fire     NaN  39      52       43       60       50   

     Speed  Generation  Legendary  
100     45           1      False  
101     60           1      False  
102     80           1      False  
103     80           1      False  
104     65           1      False

Öncesinde indek 1'den başlıyordu ve burada indeksin 100'den başlamasını sağladık. 

data3.index = range(100,900,1) anlamı, 100'den 900'e kadar 1'er artarak indeksi yeniden tanımladık.

18-ManipulatingDataFramesWithPandas-1 isimli dökümanda "data= data.set_index("#")" ifadesi ile bir sütunun indeks olmasını sağlamıştım. Aynı işlemi "data.index = data["#"]" ifadesi ile de yapabiliriz.

Data'nın bozulma ihtimaline karşı veri setimizi yeniden çağıralım.

In [7]:
data = pd.read_csv("pokemon.csv")

In [8]:
data.head()

#           Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1      Bulbasaur  Grass  Poison  45      49       49       65       65   
1  2        Ivysaur  Grass  Poison  60      62       63       80       80   
2  3       Venusaur  Grass  Poison  80      82       83      100      100   
3  4  Mega Venusaur  Grass  Poison  80     100      123      122      120   
4  5     Charmander   Fire     NaN  39      52       43       60       50   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False  
3     80           1      False  
4     65           1      False

 HIERARCHICAL INDEXING

In [9]:
data1 = data.set_index(["Type 1","Type 2"]) 
data1.head(100)  #İlk 100 sample'ı yazdırdım.

#           Name   HP  Attack  Defense  Sp. Atk  Sp. Def  \
Type 1 Type 2                                                               
Grass  Poison    1      Bulbasaur   45      49       49       65       65   
       Poison    2        Ivysaur   60      62       63       80       80   
       Poison    3       Venusaur   80      82       83      100      100   
       Poison    4  Mega Venusaur   80     100      123      122      120   
Fire   NaN       5     Charmander   39      52       43       60       50   
...            ...            ...  ...     ...      ...      ...      ...   
Poison NaN      96         Grimer   80      80       50       40       50   
       NaN      97            Muk  105     105       75       65      100   
Water  NaN      98       Shellder   30      65      100       45       25   
       Ice      99       Cloyster   50      95      180       85       45   
Ghost  Poison  100         Gastly   30      35       30      100       35   

               Speed  Generation  Legendary  
Type 1 Type 2                                
Grass  Poison     45           1      False  
       Poison     60           1      False  
       Poison     80           1      False  
       Poison     80           1      False  
Fire   NaN        65           1      False  
...              ...         ...        ...  
Poison NaN        25           1      False  
       NaN        50           1      False  
Water  NaN        40           1      False  
       Ice        70           1      False  
Ghost  Poison     80           1      False  

[100 rows x 10 columns]

Type1 ve Type2 yeni indekslerdir. İlk yazdığımız (Type 1) outer yani ilk indeks oluyor. İkinci yazdığımız (Type 2) ise inner yani 2.indeks oluyor.

Çıktıyı inceleyelim. 

Type 1'i Grass olup Type 2'si Poison olmayan bir pokemon yok. Yani bunun anlamı, böcekler (Grass) genelde zehirlidir (poison). 

Su (Water) pokemonlarının Type 2 'si yokmuş (NaN).

PIVOTING DATA FRAMES

Feature column olarak kullandığımız isimleri artık birer value ile yapıcaz demek. Yani mesela Attack ile Attack sütunun 1.satırındaki 49 değerinin yer değiştirmesi gibi. Artık 49 feature name olucak ve Attack 'da bir değer olucak.

Bunu uygulamak için bir data frame oluşturalım.

In [10]:
dic = {"treatment":["A","A","B","B"],"gender":["F","M","F","M"],"response":[10,45,5,9],"age":[15,4,72,65]}
df = pd.DataFrame(dic)
df

treatment gender  response  age
0         A      F        10   15
1         A      M        45    4
2         B      F         5   72
3         B      M         9   65

Oluşturduğumuz data frame'in kurgusu şu şekildedir:

treatment: Tedavidir.

gender: Cinsiyettir.

response: Tedaviye yanıt verebilme durumudur.

age: İnsanın yaşıdır.

In [11]:
# pivoting
df.pivot(index="treatment",columns = "gender",values="response")

gender      F   M
treatment        
A          10  45
B           5   9

Yeni indeksi "treatment" yaptık. İndeksler A ve B oldu. 

Column'lar yani feature name'ler gender'deki değerler oldu. Yani F (kadın) ve M (erkek) ŞEKLİNDE iki çeşit cinsiyet vardı ve yeni feature isimleri F ve M oldu.

Bu sütunların değerleride response sütunundaki değerler oldu.

STACKING and UNSTACKING DATAFRAME

In [12]:
df1 = df.set_index(["treatment","gender"])
df1

response  age
treatment gender               
A         F             10   15
          M             45    4
B         F              5   72
          M              9   65

treatment'ın outer ve gender'ın inner indeks olmasını sağladık. Yani ilk başta treatment'a göre diziyor ve sonra gender'a göre diziyor.

In [13]:
df1.unstack(level=0)

response    age    
treatment        A  B   A   B
gender                       
F               10  5  15  72
M               45  9   4  65

unstack() metodu, birden fazla indekse sahip data frame'lerin indeklerinin bir tanesinden kurtulmamızı sağlar.

treatment, indeks 0'dır. Ve gender, indeks 1'dir.

Burada indeks 0 'ı unstack yaptık. Yani treatment artık indeks değildir. Sadece gender indekstir.

In [14]:
df1.unstack(level=1)

response     age    
gender           F   M   F   M
treatment                     
A               10  45  15   4
B                5   9  72  65

In [15]:
df2 = df1.swaplevel(0,1)
df2

response  age
gender treatment               
F      A                10   15
M      A                45    4
F      B                 5   72
M      B                 9   65

swaplevel metodu ile indekslerin yerlerini değiştirmemiz mümkün. 

MELTING DATA FRAMES

Melting, pivot'un tam tersidir.

In [16]:
df

treatment gender  response  age
0         A      F        10   15
1         A      M        45    4
2         B      F         5   72
3         B      M         9   65

In [17]:
pd.melt(df,id_vars="treatment",value_vars=["age","response"])

treatment  variable  value
0         A       age     15
1         A       age      4
2         B       age     72
3         B       age     65
4         A  response     10
5         A  response     45
6         B  response      5
7         B  response      9

1.parametre ike data frame'i (df) melt et dedik. 2.parametre ile treatment feature'sinin sabit kalmasını sağladık. Geri kalan feature'lerin yerine default olarak variable ve value isimli iki tane feature eklenir. 3.parametre ile, age'nin değer olarak yazılması sağlanır ve age'ye karşılık gelen değerde value'ye yazılır. response içinde aynı işlem gerçekleşir. Yani variable sütun değerine df'deki sütun yazılır ve value sütununada df'deki variable sütununa yazılan sütuna karşılık gelen değer yazılır. 

Böylece data'yı melt etmiş olduk.

CATEGORICALS AND GROUPBY

In [18]:
df

treatment gender  response  age
0         A      F        10   15
1         A      M        45    4
2         B      F         5   72
3         B      M         9   65

In [19]:
df.groupby("treatment").mean()

response   age
treatment                
A              27.5   9.5
B               7.0  68.5

groupby() metodu ile bir şeye göre gruplama yapılır. Burada treatment'a göre gruplama yapılıp, ortalaması alınmıştır.

Çıktıyı inceleyelim. treatment'ı A olan response değerleri 10 ve 45'dir. (10+45)/2=27.5'dir. Görüldüğü üzere treatment'ı A ve sütunu response olan hücrede 27.5 yazmaktadır.

In [20]:
df.groupby("treatment").age.mean() 

treatment
A     9.5
B    68.5
Name: age, dtype: float64

Burada yapılan, data frame treatment'e göre gruplanır ve sonra bunun sadece age'i alınır ve bunun ortalaması alınır.

Çıktıyı inceleyelim. Treatment'i A olanların age değerleri 15 ve 4'tür. (15+4)/2=9.5'tir. 

In [21]:
df.groupby("treatment").age.max() 

treatment
A    15
B    72
Name: age, dtype: int64

Burada yapılan, data frame treatment'e göre gruplanır ve sonra her bir grubun age değeri maksimum olan alınır.

Çıktıyı inceleyelim. treatment'i A olan age değerleri 15 ve 4 'tür. Bunlardan en büyüğü olan 15 değeri çıktıya yansıtılır.

In [22]:
df.groupby("treatment")[["age","response"]].min()

age  response
treatment               
A            4        10
B           65         5

Birden fazla feature var. Bu feature'leri treatment'a göre grupladık ve sonra içerisinden sadece age ve response sütunlarını alıp min() metodunu çağırdık.

Yani treatment'e göre grupladık ve daha sonra age ve response için minimum değerlerini bulduk.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   treatment  4 non-null      object
 1   gender     4 non-null      object
 2   response   4 non-null      int64 
 3   age        4 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 256.0+ bytes


Yukarıdaki kod ile data frame hakkında bilgi edinmemiz mümkün.